### DISCUSSION TITLE
Render Lottie to UIImage with .coreAnimation

### DISCUSSION THREAD

**Discussion by artsimonyan23** [airbnb/lottie-ios#1769] [Q&A] [answered]
Render Lottie to the UIImage is not working correctly when I use the ".coreAnimation" rendering engine. 

let view: LottieAnimationView = LottieAnimationView(name: "Any Lottie")
let renderer = UIGraphicsImageRenderer(bounds: view.bounds)
let image = renderer.image { rendererContext in        
       view.layer.render(in: rendererContext.cgContext)
}

When the rendering engine is .mainThread, the image is correct. On case .coreAnimation all Lottie  components' x and y coordinates are (0, 0).

**Comment by calda** [comment, marked as answer]
This is a known problem with the Core Animation rendering engine.

The CA engine animates by playing `CAAnimation`s. These animations don't actually affect the appearance of the animation view's `layer` -- they only affect the appearance of the view's `layer.presentation()`. Because of this, the view appears blank when rendered using `view.layer.render(in:)`.

There are two possible workarounds:
 1. Render the image using `view.layer.presentation()?.render(in:)` instead. The presentation layer is only available if the view is currently being rendered on-screen in a `UIWindow`. This is probably not a problem in most application use cases, but can be tricky in headless use cases like snapshot testing (we've made this work though, [here's an example](https://github.com/airbnb/lottie-ios/blob/master/Tests/Utils/Snapshotting%2BpresentationLayer.swift#L40)).
 2. Use the Main Thread rendering engine instead, in cases where you expect to render the view into an image.

I believe this is a fundamental limitation from using `CAAnimation`s. If anyone is aware of an implementation workaround that makes this work automatically, I'd love to hear it.
